homeassistant/components/sensor/thermoworks_smoke.py

```diff
@@ -0,0 +1,178 @@
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import TEMP_FAHRENHEIT, CONF_EMAIL, CONF_PASSWORD,\
+    CONF_MONITORED_CONDITIONS, CONF_EXCLUDE, ATTR_BATTERY_LEVEL
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['thermoworks_smoke==0.1.6', 'stringcase==1.2.0']
+
+_LOGGER = logging.getLogger(__name__)
+
+PROBE_1 = 'probe1'
+PROBE_2 = 'probe2'
+PROBE_1_MIN = 'probe1_min'
+PROBE_1_MAX = 'probe1_max'
+PROBE_2_MIN = 'probe2_min'
+PROBE_2_MAX = 'probe2_max'
+BATTERY_LEVEL = 'battery'
+FIRMWARE = 'firmware'
+
+SERIAL_REGEX = '^(?:[0-9a-fA-F]{2}:){5}[0-9a-fA-F]{2}$'
+
+# map types to labels
+SENSOR_TYPES = {
+    PROBE_1: 'Probe 1',
+    PROBE_2: 'Probe 2',
+    PROBE_1_MIN: 'Probe 1 Min',
+    PROBE_1_MAX: 'Probe 1 Max',
+    PROBE_2_MIN: 'Probe 2 Min',
+    PROBE_2_MAX: 'Probe 2 Max',
+}
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_EMAIL): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_MONITORED_CONDITIONS, default=[PROBE_1, PROBE_2]):
+        vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+    vol.Optional(CONF_EXCLUDE, default=[]):
+        vol.All(cv.ensure_list, [cv.matches_regex(SERIAL_REGEX)]),
+})
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the thermoworks sensor."""
+    import thermoworks_smoke
+    from requests.exceptions import HTTPError
+
+    email = config[CONF_EMAIL]
+    password = config[CONF_PASSWORD]
+    monitored_variables = config[CONF_MONITORED_CONDITIONS]
+    excluded = config[CONF_EXCLUDE]
+
+    try:
+        mgr = thermoworks_smoke.initialize_app(email, password, True, excluded)
+
+        # list of sensor devices
+        dev = []
+
+        # get list of registered devices
+        for serial in mgr.serials():
+            for variable in monitored_variables:
+                dev.append(ThermoworksSmokeSensor(variable, serial, mgr))
+
+        add_entities(dev, True)
+    except HTTPError as error:
+        msg = "{}".format(error.strerror)
+        if 'EMAIL_NOT_FOUND' in msg or \
+                'INVALID_PASSWORD' in msg:
+            _LOGGER.error("Invalid email and password combination")
+        else:
+            _LOGGER.error(msg)
+
+
+class ThermoworksSmokeSensor(Entity):
+    """Implementation of a thermoworks smoke sensor."""
+
+    def __init__(self, sensor_type, serial, mgr):
+        """Initialize the sensor."""
+        self._name = "{name} {sensor}".format(
+            name=mgr.name(serial), sensor=SENSOR_TYPES[sensor_type])
+        self.type = sensor_type
+        self._state = None
+        self._attributes = {}
+        self._unit_of_measurement = TEMP_FAHRENHEIT
+        self._unique_id = "{serial}-{type}".format(
+            serial=serial, type=sensor_type)
+        self.serial = serial
+        self.mgr = mgr
+        self.firmware_version = ''
```

Comment:
Please remove this.

Reply:
What? That's where everything is initialized.

Reply:
The comment is only directed at the line at the end of the code box, in this case that's line 101. This is how github code reviews work.

Reply:
got it... they should let you highlight specific lines! Reverted to some code I had briefly excluding the firmware key from attrs.